### PR TITLE
Add support for TLS certificate rotation

### PR DESCRIPTION
**Pull request by DengYiping**
## What's in this PR:
 - Use OpenSSL's `set_servername_callback` to implement certificate re-load for actix
 - Caching the SSL context in memory so that we don't load certificates whenever we have a new connection. Instead, a TTL mechanism controls the re-load from disk

Idea comes from @PaulFurtado . Thanks for suggesting this change. 

### Why is it useful:
Certificate reload is important to enable users to swap certificates while the server is running. This is particularly useful for HTTPS endpoints because usually it is handled by a certificate manager (like in Kubernetes), and the certificate have a very short TTL

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### CODE REVIEW

**Comment by generall** [comment]
Hi @DengYiping, thanks for your contribution! That is some great stuff, as always!
Our dev team is partially out of the office for the holidays, so review might be a bit delayed.

Out of curiosity, are you building a Fort Knox there? ;)

**Comment by DengYiping** [comment]
> Out of curiosity, are you building a Fort Knox there? ;)

Maybe 😉 . I guess vectors in Qdrant are like gold. We need to keep it safe, and it can always benefit from an extra layer of security.

File: config/config.yaml [hunk 1]
```diff
@@ -192,3 +192,9 @@ tls:
   #
   # Required if cluster.p2p.enable_tls is true.
   ca_cert: ./tls/cacert.pem
+
+  # TTL to re-load certificate from disk. Useful for certificate rotations,
+  # Only works for HTTPS endpoints, gRPC endpoints (including intra-cluster communication)
+  # doesn't support certificate re-load
+  cert_ttl: 3600
```
Comment (timvisee) [inline config/config.yaml:199]: This should mention it is in seconds.

Also, this value is different than

```rust
// Default one
60 * 60 * 24
```

in `settings.rs`. I think we should pick one.

File: src/settings.rs [hunk 2]
```diff
@@ -192,6 +194,11 @@ fn default_message_timeout_tics() -> u64 {
     10
 }
 
+fn default_tls_cert_ttl() -> u64 {
```
Comment (timvisee) [inline src/settings.rs:197]: Nitpick: we can make this `const fn`.

File: src/actix/certificate_helpers.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    fn try_get_ssl_context(&self) -> Option<&SslContext> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Some(&self.ssl_context)
+        } else {
+            None
+        }
+    }
+
+    fn refresh_and_get_ssl_context(&mut self) -> Result<&SslContext, ()> {
```
Comment (timvisee) [inline src/actix/certificate_helpers.rs:47]: I think I like

```suggestion
    fn get_ssl_context_or_refresh(&mut self) -> Result<&SslContext, ()> {
```

a bit better. What do you think?

File: src/actix/certificate_helpers.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    fn refresh(&mut self) -> Result<(), ()> {
+        log::info!("Refreshing TLS certificates for actix!");
+        match build_ssl_context(&self.tls_config, self.verify_client_cert) {
+            Ok(ssl_context) => {
+                self.ssl_context = ssl_context;
+                Ok(())
+            }
+            Err(_err) => Err(()),
+        }
```
Comment (timvisee) [inline src/actix/certificate_helpers.rs:64]: This is a bit more concise:

```suggestion
        self.ssl_context =
            build_ssl_context(&self.tls_config, self.verify_client_cert).map_err(|_| ())?;
        Ok(())
```

File: src/actix/certificate_helpers.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    fn refresh_and_get_ssl_context(&mut self) -> Result<&SslContext, ()> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Ok(&self.ssl_context)
+        } else {
+            self.refresh()?;
+            Ok(&self.ssl_context)
+        }
```
Comment (timvisee) [inline src/actix/certificate_helpers.rs:53]:
```suggestion
        if self.last_updated.elapsed() >= self.refresh_interval {
            self.refresh()?;
        }
        Ok(&self.ssl_context)
```

File: src/actix/certificate_helpers.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                .map_err(|_refresh_error| {
+                    log::error!("Failed to refresh certificates!");
+                    SniError::ALERT_FATAL
+                })?;
+            ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
+                log::error!("Failed to set SSL context: {}", error_stack);
+                SniError::ALERT_FATAL
+            })?;
+        }
+        Ok(())
```
Comment (timvisee) [inline src/actix/certificate_helpers.rs:110]: Though this functions the same, it removes the `should_refresh` boolean:

```suggestion
        {
            let reader = ssl_context_holder.read();
            if let Some(ssl_context) = reader.try_get_ssl_context() {
                ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
                    log::error!("Failed to set SSL context: {}", error_stack);
                    SniError::ALERT_FATAL
                })?;
                return Ok(());
            }
        }

        // If getting the SSL context failed, try to get it again with refreshing
        let mut writer = ssl_context_holder.write();
        let ssl_context = writer
            .refresh_and_get_ssl_context()
            .map_err(|_refresh_error| {
                log::error!("Failed to refresh certificates!");
                SniError::ALERT_FATAL
            })?;
        ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
            log::error!("Failed to set SSL context: {}", error_stack);
            SniError::ALERT_FATAL
        })?;
        Ok(())
```

**Review by timvisee — CHANGES_REQUESTED** [review]
Nice work on this!

I know it is still in draft, but I took the liberty to review it as is anyway. Left some review comments. Mostly nitpicks. No real issues.

I think I'd like watching the certificate files for changes a bit better. Is there a good reason you choose to use a TTL instead?

**Comment by DengYiping** [comment]
> I think I'd like watching the certificate files for changes a bit better. Is there a good reason you choose to use a TTL instead?

I have indeed put this under some thoughts, but I think TTL approach is easier to implement, more portable and it is good enough for certificate rotation. Let me explain:

- Watch files are non-trivial to implement. For example, if you are using a network disk, you cannot rely on linux hook for file monitoring anymore. In Windows, I think the API is very different. And of course, if you are going to mount file from S3, that's not going to work as well.

- TTL based approach works well enough for certificate rotations because usually, when you rotate the certificates, the old certificate should still work for a short time period for all downstream clients or services to pick up the new certificates. (if not, people should revisit how they are rotating the certificate).

File: config/config.yaml (hunk 1 above)
Comment (DengYiping) [reply config/config.yaml:199]: Thanks, I picked 1 hour here as the default.

File: src/actix/certificate_helpers.rs (hunk 3 above)
Comment (DengYiping) [reply src/actix/certificate_helpers.rs]: Thanks for the suggestion, changed now accordingly.

File: src/actix/certificate_helpers.rs (hunk 6 above)
Comment (DengYiping) [reply src/actix/certificate_helpers.rs]: Done! Thanks for the suggestion!

**Comment by timvisee** [comment]
> I have indeed put this under some thoughts, but I think TTL approach is easier to implement, more portable and it is good enough for certificate rotation. Let me explain:
> 
>  * Watch files are non-trivial to implement. For example, if you are using a network disk, you cannot rely on linux hook for file monitoring anymore. In Windows, I think the API is very different. And of course, if you are going to mount file from S3, that's not going to work as well.
> 
>  * TTL based approach works well enough for certificate rotations because usually, when you rotate the certificates, the old certificate should still work for a short time period for all downstream clients or services to pick up the new certificates. (if not, people should revisit how they are rotating the certificate).

Thanks a lot for your explanation. That makes sense. I figured we could do both, but keeping it simple like this is good.

**Review by timvisee — APPROVED** [review]
LGTM!

@generall Do we want an option to disable the TTL, and do we like to disable it by default? I don't think that is required and it's fine as-is, but I want to ask for your opinion anyway.

**Comment by generall** [comment]
> @generall Do we want an option to disable the TTL, and do we like to disable it by default? I don't think that is required and it's fine as-is, but I want to ask for your opinion anyway.

Do I understand it right, that by default it will try to re-load the certificate every hour, but if certificate didn't change - functionally it will continue to work as before?

**Comment by timvisee** [comment]
> Do I understand it right, that by default it will try to re-load the certificate every hour, but if certificate didn't change - functionally it will continue to work as before?

@generall Correct! :heavy_check_mark:

**Comment by generall** [comment]
works for me

**Review by generall — APPROVED** [review]

**Pull request merged**